File: listings/cl-existence-oriented/lastpos.cxx
#include <iostream>
#include <vector>

void print_last_item( const std::vector<int>& items ) {
    std::cerr << "last item: " << items.back( ) << std::endl;
}

void safe_print_last_item( const std::vector<int>& items ) {
    if ( !items.empty( ) ) {
        std::cerr << "last item: " << items.back( ) << std::endl;
    }
}

int main( ) {
    const std::vector<int> empty_items;
    const std::vector<int> items { 42 };

    safe_print_last_item( items );
    safe_print_last_item( empty_items );
    print_last_item( items );
    print_last_item( empty_items );  // undefined behavior
}

File: listings/cl-existence-oriented/lastpos.rs
use std::io::{stderr, Write};

fn main() {
    let empty_items = vec![];
    let items = vec![42];
    unwrap_print_last_item(items);
    unwrap_print_last_item(empty_items);
}

fn unwrap_print_last_item(items: Vec<i32>) {
    if let Some(last_item) = items.last() {
        writeln!(stderr(), "last item: {}", last_item).ok();
    }
}
